File: JavaCheatSheet/src/main/java/com/github/jameswdelancey/javaCheatSheet/Part5.java
package com.github.jameswdelancey.javaCheatSheet;

class Part5 {
    public static void main(String[] args) {
        int n = 42;
        System.out.println("F(" + n + ") = " + Part5Lib.fib(n));
    }
}

class Part5Lib {
    public static void main(String[] args) {
        int n = 42;
        System.out.println("F(" + n + ") = " + fib2(n));
    }

    public static int fib(int n) {
        if (n == 0) {
            return 0;
        } else if (n == 1) {
            return 1;
        } else {
            return fib(n - 1) + fib(n - 2);
        }
    }

    public static int fib2(int n) {
        if (n == 0) {
            return 0;
        } else if (n == 1) {
            return 1;
        } else {
            int x = 1;
            int a = 0;
            int b = 1;
            while (true) {
                int next = a + b;
                a = b;
                b = next;
                x++;
                if (x < n) {
                    continue;
                }
                break;
            }
            return b;
        }
    }
}

File: rustCheatSheet/src/part5.rs
#[test]
fn main() {
    let n = 42;
    println!("F({}) = {}", n, fib(n));
}

#[test]
fn main2() {
    let n = 42;
    println!("F({}) = {}", n, fib2(n));
}

fn fib(n: u32) -> u32 {
    if n == 0 {
        0
    } else if n == 1 {
        1
    } else {
        fib(n - 1) + fib(n - 2)
    }
}

fn fib2(n: u32) -> u32 {
    if n == 0 {
        0
    } else if n == 1 {
        1
    } else {
        let mut x = 1;
        let mut a = 0;
        let mut b = 1;
        loop {
            let next = a + b;
            a = b;
            b = next;
            x += 1;
            if x < n {
                continue;
            }
            break;
        }
        b
    }
}
